mlflow/llama_index/pyfunc_wrapper.py

```diff
@@ -0,0 +1,150 @@
+
+    @staticmethod
+    def _convert_inferred_chat_message_history_to_chat_message_objects(data: Dict) -> Dict:
+        for name in _CHAT_MESSAGE_HISTORY_PARAMETER_NAMES:
+            if inferred_message_history := data.get(name):
+                if isinstance(inferred_message_history, list):
+                    for i, message in enumerate(inferred_message_history):
+                        if isinstance(message, dict):
+                            data[name][i] = ChatMessage(**message)
+
+            return data
+
+        _logger.warning(
+            f"Could not find any chat message history in the input data. "
+            f"Expected one of {_CHAT_MESSAGE_HISTORY_PARAMETER_NAMES}."
+        )
+        return data
+
+    def _format_predict_input(self, data) -> Union[str, Dict, List]:
+        data = _convert_llm_input_data(data)
+
+        if isinstance(data, str):
+            return data
+        elif isinstance(data, dict):
+            return self._convert_inferred_chat_message_history_to_chat_message_objects(data)
+        elif isinstance(data, list):
+            prediction_input = [self._format_predict_input(d) for d in data]
+            return prediction_input if len(prediction_input) > 1 else prediction_input[0]
+        else:
+            raise ValueError(
+                f"Unsupported input type: {type(data)}. It must be one of "
+                "[str, dict, list, numpy.ndarray, pandas.DataFrame]"
+            )
+
+
+class QueryEngineWrapper(_LlamaIndexModelWrapperBase):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.engine_type = QUERY_ENGINE_NAME
+
+    def _build_engine_method(self) -> Callable:
+        return self.index.as_query_engine(**self.model_config).query
+
+    def _format_predict_input(self, data) -> QueryBundle:
+        """Convert pyfunc input to a QueryBundle."""
+        data = _convert_llm_input_data(data)
+
+        if isinstance(data, str):
+            return QueryBundle.from_dict({"query_str": data})
+        elif isinstance(data, dict):
+            try:
+                return QueryBundle.from_dict(data)
+            except KeyError:
+                raise ValueError(
+                    "The input dictionary did not have the correct schema. It must support the "
+                    "supported in a llama_index QueryBundle: "
+                    "['query_str', 'image_path', 'custom_embedding_strs', 'embedding']"
+                )
+        elif isinstance(data, list):
+            prediction_input = [self._format_predict_input(d) for d in data]
+            return prediction_input if len(prediction_input) > 1 else prediction_input[0]
+        else:
+            raise ValueError(
+                f"Unsupported input type: {type(data)}. It must be one of "
+                "[str, dict, list, numpy.ndarray, pandas.DataFrame]"
+            )
+
+
+class RetrieverEngineWrapper(QueryEngineWrapper):
```

Comment:
Can we avoid subclassing query engine here? Using inheritance for reusing code is not a good practice as it causes unexpected coupling, for example,
```
retriever_wrapper = create_engine_wrapper(index, engine_type="retriever")
isinstance(retriever_wrapper, QueryEngineWrapper)
```
This will return `True`, while retriever is not a kind of query engine. I would instead define helper function like `format_query` outside the class, and call it from both query engine and retriever engine.

Reply:
Good to know. I was using subclassing solely for that purpose. Thanks

Reply:
Great to know, thanks!